src/table.js

```diff
@@ -684,10 +762,102 @@ export function __table(source, operations) {
       }))
     );
   }
-  if (primitive) source = source.map((d) => d.value);
   if (source !== input) {
     if (schema) source.schema = schema;
     if (columns) source.columns = columns;
   }
   return source;
 }
+
+function coerceRow(object, types, schema) {
+  const coerced = {};
+  for (const col of schema) {
+    const type = types.get(col.name);
+    const value = object[col.name];
+    coerced[col.name] = type === "raw" ? value : coerceToType(value, type);
+  }
+  return coerced;
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0,
+    buffer: 0
+  };
+}
+
+// We need to show *all* keys present in the array of Objects
+function getAllKeys(rows) {
+  const keys = new Set();
+  for (const row of rows) {
+    // avoid crash if row is null or undefined
+    if (row) {
+      // only enumerable properties
+      for (const key in row) {
+        // only own properties
+        if (Object.prototype.hasOwnProperty.call(row, key)) {
+          // unique properties, in the order they appear
+          keys.add(key);
+        }
+      }
+    }
+  }
+  return Array.from(keys);
+}
+
+export function inferSchema(source, columns = getAllKeys(source)) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  const typeCounts = {};
+  for (const d of sample) {
+    for (const col of columns) {
+      if (!typeCounts[col]) typeCounts[col] = initKey();
+      const type = typeof d[col];
+      const value = type === "string" ? d[col].trim() : d[col];
+      if (type !== "string") {
+        if (Array.isArray(value)) ++typeCounts[col].array;
+        else if (value instanceof Date) ++typeCounts[col].date;
+        else if (value instanceof ArrayBuffer) ++typeCounts[col].buffer;
+        else if (type === "number") {
+          ++typeCounts[col].number;
+          if (Number.isInteger(+value)) ++typeCounts[col].integer;
+        }
+        // bigint, boolean, or object
+        else if (type in typeCounts[col]) ++typeCounts[col][type];
+        else if (value !== null && value !== undefined) ++typeCounts[col].other;
+      } else {
+        if (value.toLowerCase() === "true" || value.toLowerCase() === "false") {
+          ++typeCounts[col].string;
+          ++typeCounts[col].boolean;
+        } else if (value && !isNaN(value)) {
+          ++typeCounts[col].number;
+          if (Number.isInteger(+value)) ++typeCounts[col].integer;
+        } else if (/^\d+n$/.test(value)) ++typeCounts[col].bigint;
+        else if (DATE_TEST.test(value)) ++typeCounts[col].date;
+        else if (value) ++typeCounts[col].string;
+      }
+    }
+  }
+  for (const col in typeCounts) {
+    let type = greatest(Object.keys(typeCounts[col]), d => typeCounts[col][d]);
+    const numDefined = sample.filter(d => !(d[col] == null || d[col] === "")).length;
```

Comment:
This is testing for the empty string, but not the trimmed empty string i.e. the all-whitespace string. Hence this is inconsistent with the tests above. Maybe what we should do instead is count the number of non-missing values in the iteration above, so that we don’t have to repeat a similar test here for what is missing.

Reply:
I really like the simplicity of creating and incrementing a `numDefined` property, but when I tried implementing it, it felt like it started to get complicated to figure out where/when to increment it. So I added a new check that takes the trimmed string into account... let me know what you think?